src/hist/basehist.py

```diff
@@ -150,20 +161,140 @@ def project(self, *args: Union[int, str]):
-                for index, axis in enumerate(self.axes):
-                    if name == axis.name:
-                        indices += (index,)
-                        break
-                else:
-                    raise ValueError("The axis names could not be found")
+                if isinstance(
+                    name, str
+                ):  # ToDo: redundant line - mypy cannot realized it
+                    indices += (self._name_to_index(name),)
 
             return super().project(*indices)
 
         else:
             raise TypeError(
-                f"Only projections by indices and names are supported for {self.__class__.__name__}"
+                f"Only projection by indices or names is supported for {self.__class__.__name__}"
+            )
+
+    def fill(self, *args, **kwargs):
+        """
+            Insert data into the histogram using names and indices return \
+            a Hist object.
+        """
+
+        if len(args) and not len(kwargs):
+            return super().fill(*args)
+
+        elif (
+            isinstance(kwargs, dict)
+            and all(isinstance(k, str) for k in kwargs.keys())
+            and not len(args)
+        ):
+            indices: dict = {}
+            for n, v in kwargs.items():
+                indices[self._name_to_index(n)] = v
+
+            lst = sorted(indices.items(), key=lambda item: item[0])
+            nd = np.asarray(lst, dtype=object)
+            data = nd.ravel()[1::2]
+            return super().fill(*data)
+
+        else:
+            raise TypeError(
+                f"Only dict with keys of int or str is supported for {self.__class__.__name__}"
             )
 
+    def _loc_shortcut(self, x):
+        """
+            Convert some specific indices to location.
+        """
+
+        if isinstance(x, slice):
+            return slice(
+                self._loc_shortcut(x.start),
+                self._loc_shortcut(x.stop),
+                self._step_shortcut(x.step),
+            )
+        elif isinstance(x, complex):
+            if x.real % 1 != 0:
+                raise ValueError("The real part should be an integer")
+            else:
+                return bh.loc(x.imag, int(x.real))
+        elif isinstance(x, str):
+            return bh.loc(x)
+        else:
+            return x
+
+    def _step_shortcut(self, x):
+        """
+            Convert some specific indices to step.
+        """
+
+        if isinstance(x, complex):
+            if x.real != 0:
+                raise ValueError("The step should not have real part")
+            elif x.imag % 1 != 0:
+                raise ValueError("The imaginary part should be an integer")
+            else:
+                return bh.rebin(int(x.imag))
+        else:
+            return x
+
+    def __getitem__(self, index):
+        """
+            Get histogram item.
+        """
+
+        if isinstance(index, dict):
+
+            if all(isinstance(k, int) for k in index.keys()):
```

Comment:
We should allow any mix, no reason to be extra restrictive with `all`. Just loop over the keys, if string convert to int, if int just use it, and if anything else, throw the TypeError. The only extra check I'd do would be to make sure we don't add the same int twice, as `{'x': ..., 0: ...}` is valid but should actually throw an error since it's listing the same axes twice.

Reply:
> We should allow any mix

I have thought about this. But don't you think it's odd to use a mixed style?

Reply:
Yes, but it's not something we should enforce. If it makes sense (ie, axis numbers and axes names are not overlapping), we should allow it. We should have an error if the same number is added "twice" (once as a name, once as a number), though.

Reply:
We probably will not show mixed styles in the examples, for example ;)

Reply:
After you do this, I can look into the xfails.

Reply:
And eigenvalue error is becoming more and more rampant.

Reply:
We will solve it properly in the near future.

Reply:
@henryiii Okay, You can have a look. I did not add more codes on duplicate access (access the same axis), but I believe underlying BH would throw an error if we do that. Feel free to add a test related to this issue :)